modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlPropertiesTest.java

```diff
@@ -0,0 +1,233 @@
+            ScrollPane.class,
+            Separator.class,
+            SeparatorMenuItem.class,
+            Slider.class,
+            Spinner.class,
+            SplitPane.class,
+            TableRow.class,
+            TableView.class,
+            TabPane.class,
+            TextArea.class,
+            TextField.class,
+            TextFieldTreeCell.class,
+            TextFieldTreeTableCell.class,
+            TitledPane.class,
+            ToggleButton.class,
+            TreeCell.class,
+            TreeTableCell.class,
+            TreeTableRow.class,
+            TreeTableView.class,
+            WebColorField.class
+        };
+    }
+
+    /**
+     * Tests for missing final keyword in Control properties and their accessors.
+     */
+    @Test
+    public void testMissingFinalMethods() {
+        for (Class c: allControlClasses()) {
+            check(c);
+        }
+    }
+
+    private void check(Class cls) {
+        Method[] methods = cls.getMethods();
+
+        HashMap<String, Method> h = new HashMap<>();
+        for (Method m: methods) {
+            h.put(m.getName(), m);
+        }
+
+        for (Method m: methods) {
+            String name = m.getName();
+            if (name.endsWith("Property")) {
+                int mod = m.getModifiers();
+                if (Modifier.isPublic(mod)) {
+                    if (FAIL_FAST) {
+                        Assert.assertTrue(err(m, "is not final"), Modifier.isFinal(mod));
+                    } else {
+                        if (!Modifier.isFinal(mod)) {
+                            System.err.println(err(m, "is not final"));
+                        }
+                    }
+                }
+
+                String propName = name.substring(0, name.length() - "Property".length());
+                check(h, propName, "get", 0);
+                check(h, propName, "set", 1);
+                check(h, propName, "is", 0);
+            }
+        }
+    }
+
+    private void check(HashMap<String, Method> map, String name, String prefix, int numArgs) {
+        StringBuilder sb = new StringBuilder(64);
+        sb.append(prefix);
+        sb.append(Character.toUpperCase(name.charAt(0)));
+        sb.append(name, 1, name.length());
+
+        String s = sb.toString();
+        Method m = map.get(s);
+        if (m != null) {
+            if (m.getParameterCount() == numArgs) {
+                int mod = m.getModifiers();
+                if (Modifier.isPublic(mod)) {
+                    if (FAIL_FAST) {
+                        Assert.assertTrue(err(m, "is not final"), Modifier.isFinal(mod));
+                    } else {
+                        if (!Modifier.isFinal(mod)) {
+                            System.err.println(err(m, "is not final"));
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private String err(Method method, String message) {
+        return method + " " + message;
+    }
```

Comment:
(This comment didn't appear in the review for some reason.)

I think that the code could be a bit more clean. I took a stab at it:

```java
    Set<Class<?>> CLASSES = Set.of(/*...*/);

    private void check(Class<?> cls) {
        Map<String, Method> methods = Stream.of(cls.getMethods()).collect(Collectors.toMap(Method::getName, m -> m)); // the loop is also fine
        
        for (var method : methods.values()) {
            String name = method.getName();
            if (name.endsWith("Property")) {
                checkModifiers(method);

                String propName = name.substring(0, name.length() - "Property".length());
                check(methods, propName, "get", 0);
                check(methods, propName, "set", 1);
                check(methods, propName, "is", 0);
            }
        }
    }

    private void check(Map<String, Method> methods, String name, String prefix, int numArgs) {
        String methodName = new StringJoiner("", prefix, name.substring(0, 1).toUpperCase() + name.substring(1)).toString();

        Method m = methods.get(methodName);
        if (m != null && m.getParameterCount() == numArgs) {
            checkModifiers(m);
        }
    }

    private void checkModifiers(Method m) {
        int mod = m.getModifiers();
        if (Modifier.isPublic(mod) && !Modifier.isFinal(mod)) {
            if (FAIL_FAST) {
                throw new AssertionError(errorMessage(m));
            } else {
                System.err.println(errorMessage(m));
            }
        }
    }

    private String errorMessage(Method method) {
        return method + "is not final";
    }
```

Reply:
Thank you for suggestions!

Frankly, I don't see much value in rewriting of this method.  A more interesting problem is how to enumerate all descendants of the Control class  using the package name, in such a way that will reliably work in an IDE, from a JAR, or in any test framework (jtreg or otherwise).

Reply:
> edit: about the test itself - I see little difference between the two version, but will gladly use your version if you feel strongly about it.

I don't, I just think it's more readable. Fine to leave as is. What I might strongly suggest is extracting the `checkModifiers` method that is used in 2 places in your version.

Reply:
done, and also fixed a stupid mistake ... method names are not unique.
many thanks!